Feature/epnp

zitongzhan opened the pull request:
https://github.com/pypose/pypose/issues/149

- Made the function a torch.nn.Module
- Batched operation. However, does not allow arbitrary batch shape (`...`) because of the limitations of svd and vmap functions
- Refactored all naming
- Added docstring

[Diff](https://github.com/pypose/pypose/compare/b5bb567588a17f711009ce4ac068e0c42e4560d7..f4fd43260d6c1ddab5d4bb296571358bacb73f3b) with original implementation. Naming are changed to all lower case but correspondence are still identifiable.

wang-chen commented on pypose/module/pnp.py:
```diff
@@ -0,0 +1,483 @@
+
+class EfficientPnP(torch.nn.Module):
+    """
+        EPnP Solver - a non-iterative O(n) solution to the PnP problem.
+        as described in:
+
+        Francesc Moreno-Noguer, Vincent Lepetit, Pascal Fua.
+        Accurate Non-Iterative O(n) Solution to the PnP Problem.
+        In Proceedings of ICCV, 2007.
+        source: https://github.com/cvlab-epfl/EPnP
+
+        Args:
+            gauss_newton (bool): Use Gauss-Newton to refine the solution.
+            naive_ctrl_pts (bool): Use naive control points selection method.
+
+        Note:
+            The Gauss-Newton optimization step isn't exactly consistent with the implementation in the original paper's
+            implementation, but it relies on the same optimization objective function.
```
add example and algorithm.

wang-chen commented on pypose/module/pnp.py:
```diff
@@ -0,0 +1,483 @@
+            The Gauss-Newton optimization step isn't exactly consistent with the implementation in the original paper's
+            implementation, but it relies on the same optimization objective function.
+    """
+    six_indices_pair = torch.tensor([(0, 1), (0, 2), (0, 3), (1, 2), (1, 3), (2, 3)], device='cpu').T
+    six_indices = [(0 * 4 + 1), (0 * 4 + 2), (0 * 4 + 3), (1 * 4 + 2), (1 * 4 + 3), (2 * 4 + 3)]
+    ten_indices_pair = torch.tensor([(0, 0),
+                                     (0, 1), (1, 1),
+                                     (0, 2), (1, 2), (2, 2),
+                                     (0, 3), (1, 3), (2, 3), (3, 3)], device='cpu').T
+    # equal mask [ True, False,  True, False, False,  True, False, False, False,  True]
+    multiply_mask = torch.tensor([1., 2., 1., 2., 2., 1., 2., 2., 2., 1.], device='cpu')
```
move to `init()` and define as `register_buffer()` so that it can follow the device of the module.

wang-chen commented on pypose/module/pnp.py:
```diff
@@ -0,0 +1,483 @@
+    # equal mask [ True, False,  True, False, False,  True, False, False, False,  True]
+    multiply_mask = torch.tensor([1., 2., 1., 2., 2., 1., 2., 2., 2., 1.], device='cpu')
+
+    def __init__(self, gauss_newton=True, naive_ctrl_pts=False):
```
take optimizer as an argument, e.g., `LM()` or `SGD()`.

wang-chen reviewed (changes requested):
Following our meeting, better to give an API design first: showing how we want the user to use this module. Just like the example we provided in [LM optimizer](https://pypose.org/docs/main/generated/pypose.optim.LevenbergMarquardt/#pypose.optim.LevenbergMarquardt).

zitongzhan commented:
- changed to pypose bmv
- changed interface definition
- add camera base class and perspective camera
- add example usage
- remove bsvd todos, since pytorch does support bsvd

wang-chen commented on pypose/module/cameras.py:
```diff
@@ -0,0 +1,149 @@
+        Returns:
+            A tensor of shape (B, 3) where B is the batch size.
+        """
+        return pypose.Inv(self.pose)[..., :3]
```
`return self.pose.Inv().translation()`

wang-chen commented on pypose/module/cameras.py:
```diff
@@ -0,0 +1,149 @@
+        Returns:
+            points (torch.Tensor): A tensor of shape (B, N, 2) where B is the batch size.
+        """
+        img_repj = torch.matmul(points, self.intrinsics.transpose(-2, -1))
```
do you want
```python
img_repj = self.intrinsics @ points
```
or
```python
img_repj = self.intrinsics.mT @ points
```

zitongzhan replied:
I wrote `points.matmul(self.intrinsics.transpose(-2, -1))` because it would be the cleanest way of writing it. Left matmul of intrinsics has some problem with broadcasting

```
self.intrinsics.shape
PyDev console: starting.
torch.Size([1, 3, 3])
points.shape
torch.Size([1, 5, 3])
pypose.bmv(self.intrinsics.unsqueeze(-2), points)
Traceback (most recent call last):
  File "/Applications/PyCharm.app/Contents/plugins/python/helpers/pydev/_pydevd_bundle/pydevd_exec2.py", line 3, in Exec
    exec(exp, global_vars, local_vars)
  File "<input>", line 1, in <module>
  File "/Users/zitongzhan/Documents/pypose/pypose/basics/ops.py", line 34, in bmv
    return torch.matmul(mat, vec.unsqueeze(-1), out=out).squeeze_(-1)
RuntimeError: The size of tensor a (3) must match the size of tensor b (5) at non-singleton dimension 1
```

zitongzhan replied:
changing it to `img_repj = points.matmul(self.intrinsics.mT)`

wang-chen commented on pypose/module/cameras.py:
```diff
@@ -0,0 +1,149 @@
+        img_repj = self.full_projection_transform(pts_w)
+
+        error = torch.linalg.norm(img_repj - img_pts, dim=-1)
+        error = torch.mean(error, dim=-1)
```
`error = error.mean(dim=-1)`

wang-chen commented on pypose/module/cameras.py:
```diff
@@ -0,0 +1,149 @@
+        """
+        return pypose.Inv(self.pose)[..., :3]
+
+    def world_to_view_transform(self, points):
```
How about `def world2camera(self, points):`
this will follow our [naming convention](https://pypose.org/docs/preview/pr/184/convert/).

wang-chen commented on pypose/module/cameras.py:
```diff
@@ -0,0 +1,149 @@
+        """
+        return pypose.Act(self.pose.unsqueeze(-2), points)
+
+    def full_projection_transform(self, points):
```
how about `def world2pixel(self, points):`

wang-chen commented on pypose/module/cameras.py:
```diff
@@ -0,0 +1,149 @@
+        """
+        return self.projection_transform(self.world_to_view_transform(points))
+
+    def projection_transform(self, points):
```
how about `def camera2pixel(self, points):`

wang-chen commented on pypose/module/pnp.py:
```diff
@@ -0,0 +1,486 @@
+            diff_w = diff_w.reshape(*batch_shape, 16, 3)
+            diff_w = torch.sum(diff_w ** 2, dim=-1)
+
+            error = torch.abs(diff_w - diff_c)
+            error = torch.mean(error, dim=-1)
```
```
error = (diff_w - diff_c).abs().mean(dim=-1)
```

jingnanshi commented on pypose/module/cameras.py:
```diff
@@ -0,0 +1,150 @@
+        img_repj = points.matmul(self.intrinsics.mT)
+        return img_repj[..., :2] / img_repj[..., 2:]
+
+    def is_perspective(self):
```
Design: is this function really necessary? Is there a camera class other than PerspectiveCamera that is also perspective? If not, I would suggest removing this function (or at least implement it in the base class, and default to return False)

zitongzhan replied:
not really necessary. going to remove it

wang-chen commented on pypose/module/cameras.py:
```diff
@@ -0,0 +1,150 @@
+        Returns:
+            points (torch.Tensor): A tensor of shape (B, N, 3) where B is the batch size.
+        """
+        return pypose.Act(self.pose.unsqueeze(-2), points)
```
Use `self.pose.unsqueeze(-2) @ points` could be more intuitive, but do you really need `.unsqueeze(-2)`? How about `self.pose @ points`?

zitongzhan replied:
self.pose @ points doesn't work. Still need an additional dimension for broadcasting

wang-chen commented on pypose/module/cameras.py:
```diff
@@ -0,0 +1,150 @@
+            intrinsics: Optional[torch.Tensor] = None,
+    ):
+        super().__init__(pose=pose)
+        if intrinsics is not None:
+            self.intrinsics = intrinsics
+        elif self.intrinsics is None:
+            self.intrinsics = torch.eye(3)
```
similar to above, `self.pose`

wang-chen commented on pypose/module/cameras.py:
```diff
@@ -0,0 +1,150 @@
+        >>> pose = pypose.SE3([ 0.0000, -8.0000,  0.0000,  0.0000, -0.3827,  0.0000,  0.9239])
+        >>> f = 2
+        >>> img_size = (7, 7)
+        >>> projection_matrix = torch.tensor([[f, 0, img_size[0] / 2,], [0, f, img_size[1] / 2,], [0, 0, 1, ]])
+        >>> pts_w = torch.tensor([[ 2.8284,  8.0000,  0.0000], [ 2.1213,  8.0000,  0.7071], [ 0.7071,  9.0000,  0.7071], [ 0.7071,  8.0000,  0.7071], [ 5.6569, 13.0000, -1.4142]])
```
line too long, limit to 97 characters.

wang-chen commented on test/module/test_pnp.py:
```diff
@@ -0,0 +1,269 @@
+        print("OpenCV EPnP solution, rmse of t:", rmse_t(solution_ref['T'], gt_t))
+
+
+# CPnP: a consistent PnP solver
+# Inputs: s - a 3×n matrix whose i-th column is the coordinates (in the world frame) of the i-th 3D point
+# Psens_2D - a 2×n matrix whose i-th column is the coordinates of the 2D projection of the i-th 3D point
+# fx, fy, u0, v0 - intrinsics of the camera, corresponding to the intrinsic matrix K=[fx 0 u0;0 fy v0;0 0 1]
+
+# Outputs: R - the estimate of the rotation matrix in the first step
+# t - the estimate of the translation vector in the first step
+# R_GN - the refined estimate of the rotation matrix with Gauss-Newton iterations
+# t_GN - the refined estimate of the translation vector with Gauss-Newton iterations
+# Copyright <2022> <Guangyang Zeng, Shiyu Chen, Biqiang Mu, Guodong Shi, Junfeng Wu>
+# Guangyang Zeng, SLAMLab-CUHKSZ, September 2022
+# zengguangyang@cuhk.edu.cn, https://github.com/SLAMLab-CUHKSZ
+# paper link: https://arxiv.org/abs/2209.05824
```
remove unnecessary lines

wang-chen commented on test/module/test_pnp.py:
```diff
@@ -0,0 +1,269 @@
+    epnp_fixture.test_epnp_10pts()
+    epnp_fixture.test_epnp_random()
+
+    # cpnp_fixture = TestCPnP()
+    # cpnp_fixture.test_cpnp_random()
```
remove

wang-chen commented on test/module/test_pnp.py:
```diff
@@ -0,0 +1,269 @@
+import torch
+import urllib
+
+import logging
+
+logger = logging.getLogger(__name__)
```
what is this purpose?
remove if unnecessary.

wang-chen commented on test/module/test_pnp.py:
```diff
@@ -0,0 +1,269 @@
+import io
+import os
+import pypose as pp
+import torch
+import urllib
```
```
import io, os, urllib
import torch, pypose as pp
```
Pyramid imports

wang-chen commented on pypose/module/pnp.py:
```diff
@@ -0,0 +1,493 @@
+class EPnP(torch.nn.Module):
+    r"""
+        EPnP Solver - a non-iterative O(n) solution to the PnP problem.
+        as described in:
+
+        Francesc Moreno-Noguer, Vincent Lepetit, Pascal Fua.
+        Accurate Non-Iterative O(n) Solution to the PnP Problem.
+        In Proceedings of ICCV, 2007.
+        source: https://github.com/cvlab-epfl/EPnP
```
move citation to end of doc, Use `Note:` when necessary. refer to [LQR](https://pypose.org/docs/main/generated/pypose.module.LQR)

wang-chen commented on pypose/module/pnp.py:
```diff
@@ -0,0 +1,493 @@
+        source: https://github.com/cvlab-epfl/EPnP
+
+        Args:
+            refinement_optimizer (Optional[torch.optim.Optimizer]): Optimizer class to refine the solution. Set to None to disable refinement.
+            naive_ctrl_pts (bool): Use naive control points selection method.
```
It seems that doc is not updated.
Too long. each line limit to 97 characters.

wang-chen commented on pypose/module/pnp.py:
```diff
@@ -0,0 +1,493 @@
+
+            return error
+
+    def __init__(self, naive_ctrl_pts=False, refinement_optimizer=GaussNewton):
```
Pay attention to the naming convention, how about 'optimizer'?
I will discourage `naive_ctrl_pts`, do you have better name?
Different optimizers need different arguments, better to pass optimizer variables instead of optimizer class name.

zitongzhan replied:
It's not possible to create the optimizer object before PnP begins, because the optimizer's constructor takes the `BetasOptimizationObjective` as input. BetasOptimizationObjective's constructor takes an intermediate solution `beta` as input. 

as shown in these two lines
```
        objective = self.BetasOptimizationObjective(solutions['beta'] * solutions['scale'].unsqueeze(-1))
        gn = self.refinement_optimizer(objective)
```

zitongzhan commented on test/module/test_pnp.py:
```diff
@@ -0,0 +1,269 @@
+    assert os.path.exists(tmp_path), 'download testing mat failed'
+    return tmp_path
+
+sample_binary = b'PK\x03\x04\x00\x00\x08\x08\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x14\x00\x0e\x00epfl_sample/data.pklFB\n\x00ZZZZZZZZZZ\x80\x02}q\x00(X\x06\x00\x00\x00camPtsq\x01ctorch._utils\n_rebuild_tensor_v2\nq\x02((X\x07\x00\x00\x00storageq\x03ctorch\nDoubleStorage\nq\x04X\x01\x00\x00\x000q\x05X\x03\x00\x00\x00cpuq\x06K\x96tq\x07QK\x00K2K\x03K\x01\x87q\x08K ...
```
It's a Pytorch compressed pth file that is hard to be defined directly in code

wang-chen commented on pypose/module/pnp.py:
```diff
@@ -0,0 +1,489 @@
+            lietensor.SE3Type: estimated pose for the camera
+        """
+        # shape checking
+        batch_shape = points.shape[:-2]
+        assert pixels.shape[:-2] == batch_shape
+        assert intrinsics.shape[:-2] == batch_shape
```
batch dimension is unnecessary to be the same, they can be broadcast-able, e.g., `(1,2,3)` with `(2,1,1)` is also fine.

wang-chen commented on pypose/module/pnp.py:
```diff
@@ -0,0 +1,489 @@
+import torch
+import pypose
+from pypose.optim import GaussNewton
```
we are developing `pypose`, we don't import `pypose` inside `pypose` modules. use relative imports to avoid recursive imports.

wang-chen commented on pypose/module/pnp.py:
```diff
@@ -0,0 +1,489 @@
+        return solution
+
+    @staticmethod
+    def naive_control_points(points):
+        """
+        Select four control points, used to express world coordinates of the object points. This is a naive
+        implementation that corresponds to the original paper.
+
+        Args:
+            points: 3D object points, shape (..., n, 3)
+        Returns:
+            control points, shape (..., 4, 3)
+        """
+        batch_shape = points.shape[:-2]
+        control_pts = torch.eye(3, dtype=points.dtype, device=points.device)
+        # last control point is the origin
+        control_pts = torch.cat((control_pts, torch.zeros(1, 3, dtype=points.dtype, device=points.device)), dim=0)
+        control_pts = control_pts.reshape((1,) * len(batch_shape) + (4, 3)).repeat(*batch_shape, 1, 1)
+        return control_pts
```
it seems that this only returns identity matrix with an additional zero row?

wang-chen replied:
This should be more readable.
```
controls = torch.zeros_like(points[...,:4,:])
controls.diagonal(dim1=-2, dim2=-1).fill_(1)
return controls
```

wang-chen commented on pypose/module/pnp.py:
```diff
@@ -0,0 +1,489 @@
+            control points, shape (..., 4, 3)
+        """
+        # Select the center of mass to be the first control point
+        center = points.mean(axis=-2)
+
+        # Use distance to center to select the other three control points
+        # svd
+        centered_points = points - center.unsqueeze(-2)  # center the object points, 1 is for broadcasting
+        u, s, vh = torch.linalg.svd(torch.bmm(centered_points.mT, centered_points), full_matrices=True)
+
+        # produce points TODO: change to batch implementation
+        res = [center, ]
+        for i in range(3):
+            another_pt = center + torch.sqrt(s[..., i, None]) * vh[..., i]
+            res.append(another_pt)
+
+        return torch.stack(res, dim=-2)
```
This should work and more readable:
```python
center = points.mean(dim=-2, keepdim=True)
translated = points - center
u, s, vh = torch.linalg.svd(translated.mT @ translated)
controls = center + s.sqrt().unsqueeze(-1) * vh
return torch.cat([center, controls], dim=-2)
```

zitongzhan replied:
A transpose is needed for vh

zitongzhan commented on pypose/module/pnp.py:
```diff
@@ -184,56 +187,28 @@ def optimization_step(self, solutions, kernel_m, ctrl_pts_w, alpha, points, pixe
-            res.append(another_pt)
-
-        return torch.stack(res, dim=-2)
+    def svd_control(points):
```
This change cause singular matrix exception

zitongzhan replied:
resolved in commit below

zitongzhan commented:
The last 2 commits fixes _compute_norm_sign_scaling_factor and _generate_solution, respectively 
The commits before that simplifies the forward function

zitongzhan commented:
Summary: https://github.com/pypose/pypose/compare/42b454d..402e388

jingnanshi commented on pypose/module/pnp.py:
```diff
@@ -0,0 +1,308 @@
+        r'''
+        Args:
+            points (``torch.Tensor``): 3D object points in the world coordinates.
+                Shape (..., n, 3)
```
In the other comments, "N" is used instead of "n". Consider keeping it consistent

zitongzhan replied:
Fixed in commit below

jingnanshi reviewed (approved):
Great job! I just left a few minor comments.

zitongzhan commented on pypose/module/pnp.py:
```diff
@@ -0,0 +1,308 @@
+        The implementation is based on the paper
+
+        * Francesc Moreno-Noguer, Vincent Lepetit, and Pascal Fua, `EPnP: An Accurate O(n)
+          Solution to the PnP Problem <https://doi.org/10.1007/s11263-008-0152-6>`_,
```
This is required: the underscore was correct for link generation with sphnix
